wolfcrypt/test/test.c

```diff
@@ -20252,6 +20264,551 @@ int sshkdf_test(void)
+                 if (ret == 0)
+                     ret = wc_Sha256Final(&sha256, out);
+                 wc_Sha256Free(&sha256);
+             }
+        }
+        break;
+ #endif
+ #ifdef WOLFSSL_SHA384
+        case WC_SHA384:
+        {
+            wc_Sha384 sha384;
+             ret = wc_InitSha384(&sha384);
+             if (ret == 0) {
+                 if (in1)
+                     ret = wc_Sha384Update(&sha384, in1, in1Sz);
+                 if (ret == 0 && in2)
+                     ret = wc_Sha384Update(&sha384, in2, in2Sz);
+                 if (ret == 0 && in3)
+                     ret = wc_Sha384Update(&sha384, in3, in3Sz);
+                 if (ret == 0 && in4)
+                     ret = wc_Sha384Update(&sha384, in4, in4Sz);
+                 if (ret == 0)
+                     ret = wc_Sha384Final(&sha384, out);
+                 wc_Sha384Free(&sha384);
+             }
+        }
+        break;
+ #endif
+ #ifdef WOLFSSL_TLS13_SHA512
+        case WC_SHA512:
+        {
+            wc_Sha512 sha512;
+             ret = wc_InitSha512(&sha512);
+             if (ret == 0) {
+                 if (in1)
+                     ret = wc_Sha512Update(&sha512, in1, in1Sz);
+                 if (ret == 0 && in2)
+                     ret = wc_Sha512Update(&sha512, in2, in2Sz);
+                 if (ret == 0 && in3)
+                     ret = wc_Sha512Update(&sha512, in3, in3Sz);
+                 if (ret == 0 && in4)
+                     ret = wc_Sha512Update(&sha512, in4, in4Sz);
+                 if (ret == 0)
+                     ret = wc_Sha512Final(&sha512, out);
+                 wc_Sha512Free(&sha512);
+             }
+        }
+        break;
+#endif
+    }
+    return ret;
+}
+
+static int CompareExpected(const char* expected,
+        const byte* compare, word32 compareSz)
+{
+    byte expectedBin[WC_MAX_DIGEST_SIZE];
+    int ret;
+    word32 expectedBinSz = sizeof expectedBin;
+
+    ret = ConvertHexToBin(expected, expectedBin, &expectedBinSz,
+            NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
+
+    if (ret == 0) {
+        if (expectedBinSz != compareSz ||
+                XMEMCMP(expectedBin, compare, expectedBinSz) != 0) {
+            ret = -1;
+        }
+    }
+
+    return ret;
+}
+
+
+typedef struct {
+    byte hashAlg;
+    const char* psk;
+    const char* dhe;
+    const char* helloClientRandom;
+    const char* helloServerRandom;
+    const char* finishedClientRandom;
+    const char* finishedServerRandom;
+    /* Expected */
+    const char* clientEarlyTrafficSecret;
+    const char* earlyExporterMasterSecret;
+    const char* clientHandshakeTrafficSecret;
+    const char* serverHandshakeTrafficSecret;
+    const char* clientApplicationTrafficSecret;
+    const char* serverApplicationTrafficSecret;
+    const char* exporterMasterSecret;
+    const char* resumptionMasterSecret;
+} Tls13KdfTestVector;
+
+static const Tls13KdfTestVector tls13KdfTestVectors[] = {
```

Comment:
If these test vectors came from somewhere, please document.
If not, consider pre-hashing input.

Reply:
I'll pre-hash them. The values came from one of our CAVP test sets.